trusted_os/rpmb_fake.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2022 The Armored Witness OS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build fake_rpmb
+// +build fake_rpmb
+
+package main
+
+import (
+	"encoding/binary"
+	"errors"
+	"strconv"
+
+	"github.com/transparency-dev/armored-witness-os/rpmb"
+)
+
+const (
+	// version epoch length
+	versionLength = 4
+	// RPMB sector for OS rollback protection
+	osVersionSector = 1
+	// RPMB sector for TA rollback protection
+	taVersionSector = 2
+
+	// RPMB sector for TA use
+	taUserSector = 3
+
+	sectorLength = 256
+	memoryLength = 1024
+)
+
+type RPMB struct {
+	mem     [memoryLength]byte
+	counter uint32
+}
+
+func newRPMB(_ Card) (*RPMB, error) {
+	return &RPMB{}, nil
+}
+
+func parseVersion(s string) (version uint32, err error) {
+	v, err := strconv.Atoi(s)
+
+	if err != nil {
+		return
+	}
+
+	return uint32(v), nil
+}
+
+// expectedVersion returns the version epoch stored in a fake RPMB area.
+func (r *RPMB) expectedVersion(sector uint16) (version uint32, err error) {
+	buf := make([]byte, versionLength)
+	copy(buf, r.mem[sector*sectorLength:])
+
+	return binary.BigEndian.Uint32(buf), nil
+}
+
+// updateVersion writes a new version epoch in a fake RPMB area.
+func (r *RPMB) updateVersion(sector uint16, version uint32) (err error) {
+	buf := make([]byte, versionLength)
+	binary.BigEndian.PutUint32(buf, version)
+
+	copy(r.mem[sector*sectorLength:], buf)
+	r.counter++
+
+	return nil
+}
+
+// checkVersion verifies version information against fake RPMB stored data.
+//
+// If the passed version is older than the RPMB area information of the
+// internal eMMC an error is returned.
+//
+// If the passed version is more recent than the RPMB area information then the
+// internal eMMC is updated with it.
+func (r *RPMB) checkVersion(sector uint16, s string) (err error) {
```

Comment:
Not for this PR, but can you have a think/look at what we should store for version - is a `uint32` right or do we need something more structured (~~ where does the `s` here come from, and what would we be setting it to)?

Reply:
The version comes from [here](https://github.com/transparency-dev/armored-witness-applet/blob/516443d4a507a44f6f683b94533ef1d96504678c/api/log_entries.go#L37), so I think that means we do indeed need to serialize the structured [Version data](https://github.com/coreos/go-semver/blob/6328231d806b2e2199c42ecc055a16ddcc2c6121/semver/semver.go#L27).

Noted to followup on this.